### DISCUSSION TITLE
Dynamic Power Limiter - unplanned behavior? Bug or my fault?

### DISCUSSION THREAD

**Discussion by OlSteph** [hoylabs/OpenDTU-OnBattery#256] [Q&A] [answered]
Hello,

i have the DTUonBattery with 2 inverters - one HM-600 is permamently at 100% and the 2nd Inverter HM-1500 is set as to limit.
No Battery is connected (i have no Battery System yet - maybe later)
Here are my Settings:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/8fef3248-9306-41a2-9ead-725589bf3327)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/48f0ccd6-d56e-429b-aa0e-f52f6839bbb0)

The Problem is - sometimes - like today with the washing machine heating and the Limit goes to 100%, after the heating the limit stays there and after a "long" time the limit is set correctly, but to Late.

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/33159579-6e84-4f7f-8df3-7324a04fdb52)

When it is not my mistake, how can i "debug" od analyze this to maybe help to fix this.
If this is an Bug / Error i woult open a Issue.
Thank You

**Comment by spcqike** [comment]
how does your opendtu gets the reading if your mains meter? have you connected boths inverters to the same DTU? 

if this happens again, you can check and monitor the konsole in webgui
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/e7cdb1ca-4fc1-45ea-9749-4e7ba70991dc)

there should be total power readings
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/8243b618-1a92-467d-8f91-f93d128aaa75)

and power limiter readings
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/6fd26408-8965-4708-8b5b-aa113639ea21)

**Reply by OlSteph** [reply]
The DTU has connected both inverters - if needed i could use/test with 2 different DTUs (same hardware available hiere - not flashed yet)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/3b977659-3b74-40a8-aa64-0174f6958e50)

The DTU gehts power meter readings from my Shelly Pro 3EM
(LAN Connected - no Password set because of new Digest-Authentication not yet available.)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/22375336-db67-4b21-9288-a8fbb3a34a87)

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/c5200873-f0f9-4d2e-9c2d-ff72312ca8d8)

Thanks for the Monitor tipps - i will check the Konsole and report here.

**Reply by spcqike** [reply]
can you see how frequent the total power reading (powermeterclass) is updated? it should be updated every 10s, unless you changed it in your config.

also, yesterday i changed my setting to yours. so i switched off solar-passtrough. this morning, my inverter was stopped from DTU.
(i have a HM-600 only, without battery. battery voltages and SoC set to 0, and, of course, other power limits than you :D) 

for me, dynamic limiting without a battery only worksflawlessly, if i have solar passtrough enabled. but, other than you, my problem is that it stops the inverter at all. 

maybe you can try this, too.

**Reply by OlSteph** [reply]
At the moment it is working right, i will let the konsole open and check later - the situation ws eralier today but i wasn't on the pc at that time.
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/586d3027-27b6-49f3-888e-aa085b606cba)

here is an actual log
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/e1665850-2810-4aee-9c9a-3ef0e705272b)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/f86818eb-9cf2-46db-ac2b-60a01540b6b2)

**Reply by OlSteph** [reply]
it happened again:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/14d2c455-b040-476a-8e4c-5ac6c0f8efe2)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/342ab342-9eb4-4935-9123-a4c6fd983760)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/b8c956b9-f247-4450-a3a0-2c9a439a6f63)

For tomorrow i will test  solar passtrough  on.
The Symptom that the inverter ist stopped sometimes comes here to, then i had to restart the inverter to get it on again.
Bit i tought that there is already an issue somewhere for that

**Reply by spcqike** [reply]
I guess this occurs because of your high hysteresis. Insist possible that your new target voltage (after your high load) is smaller than your 200W hysteresis?

The line can be found here 
https://github.com/helgeerbe/OpenDTU-OnBattery/blob/master/src/PowerLimiter.cpp#L261

You can try a smaller hysteresis. But, if it’s really the root cause, it could happen in any situation where you get a smaller load, I guess. 

Edit:
As I see all your new limit targets are >200W. Is it possible that you can’t dynamically limit the power to 0-200W? 

My hysteresis is 15W, so 0-15W is nothing my limiter ever would set, as my consumption is at least 150-200W 24/7. 

Edit:
Just tried set my hysteresis to 350W, so higher than my usual load. It started using the old limit value. 

With a hysteresis of 300W, it calculated new values all over the place (every 10s), even so it only changed total value by 5-20W, as my real load changed. 

I guess, the hysteresis doesn’t work as expected. 

@helgeerbe how should hysteresis work? I thought it should be a threshold to only calculate new values of the change to the last set value is greater or smaller than hysteresis. But that’s not how it works right now. Is it?

**Reply by helgeerbe** [reply]
> @helgeerbe how should hysteresis work? I thought it should be a threshold to only calculate new values of the change to the last set value is greater or smaller than hysteresis. But that’s not how it works right now. Is it?

Your are right. But there was a complete refactoring of the old code lately. So there might be an issue.

**Comment by MalteSchm** [comment]
@OlSteph 
It could be that the specific combination of values triggers a bug. You could start trying to set the upper limit to a value that the inverter can actually handle. (change the 1695W to 1500W). This is anyhow a good idea.

The other option is to set the "angestrebter Netzbezug" to 0W for testing

**Reply by OlSteph** [reply]
@spcqike 
I let solar passtrough enabled this mrning for test purpose and deactivated it later:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/b032dc6a-89b1-4375-8145-0c066854c98d)
 
@MalteSchm 
This is the Graph so far (see above) - with this settings starting around 11:00 :
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/1ed1682c-7caf-4f15-9955-33d2882d8388)
I have played around with the upper limit, because of the strange behaviours of the hysteresis - at first i have set to 1500.
(or should it be set to the total-upper limit of all inverters - which would be 2100w ?)

**Reply by spcqike** [reply]
Can you plot your power consumption? Like grid_import + pv_production? Is it possible that your power consumption wasn’t higher than 200W until 11.00? (And if it was was it long enough so that the opendtu could recongnize it? Like 10-20sec at least?) 

11.00 you changed your Netz Bezug and hysterese to 0 and 30, right? 

Afterwards the limiter worked. Or? 

IMO upper limit doesn’t matter in this case as, regardless what power limit gets calculated within the processs, at the end it’s maximum will be set to your inverters hard coded maximum (if the calculated target power is higher than what the inverter can do) 

Need to check this in the code but I think the Lower and upper limit are used before the calculation of the real limit.

**Reply by MalteSchm** [reply]
The upper limit should be set to the max power of the inverter that is controlled by the power limiter. In your case that is 1500W (max)

**Comment by OlSteph** [comment]
@spcqike 
do you mean this? above is grid import and export, the upper picture is pv prodiction (both inverters) - marked 4 are the HM-1500
the lower are detailed plots from the panels
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/c39b202a-70b3-40bb-9889-eecf0c315a6d)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/5a0a4213-ff07-4d4a-a0ae-7f02532ab8a6)

"11.00 you changed your Netz Bezug and hysterese to 0 and 30, right?" -> around 11:45

"Afterwards the limiter worked. Or?" Yes

The whole day (yesterday)
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/01fa1614-f23c-4c12-83b0-0ef55d20d377)

The little Peaks later should be "normal" - when high loads end insttantly then the few seconds hing export till the power limiting happens. (in my opinion)

**Reply by spcqike** [reply]
No I meant a combination of both, grid meter and HM-1500. so that you can see your „total energy consumption“ (in your case except your HM-600, but that’s not important) 

like at 9.00 you have a net grid „consumption“ of approx -800 and a production of ~900. which would give a real consumption of 100W. 

Just to verify that your calculated power consumption was always <200W. 

But, as your new hysteresis of 30W seems to fix the issue, I guess it doesn’t matter anymore. 

Still I think that hysteresis doesn’t work as expected.

**Reply by OlSteph** [reply]
Ok, next try:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/73e9a6ac-7be1-46fa-8d1b-854ae387b32c)

**Reply by OlSteph** [reply]
The values of my shelly 3em maybe would be more precise - but at the moment i am not able to log them properly because i dont know yet how to get seperate values from the string/array in iobroker:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/c14a5df4-490f-489d-9a0f-0b057e85c63a)
{"id":0,"a_current":0.598,"a_voltage":230.7,"a_act_power":77.2,"a_aprt_power":137.9,"a_pf":-0.69,"b_current":0.885,"b_voltage":230.6,"b_act_power":38.7,"b_aprt_power":203.9,"b_pf":-0.55,"c_current":0.989,"c_voltage":231.3,"c_act_power":186.3,"c_aprt_power":228.5,"c_pf":-0.84,"n_current":null,"total_current":2.472,"total_act_power":302.252,"total_aprt_power":570.320, "user_calibrated_phase":[]}

**Reply by spcqike** [reply]
> Ok, next try:
> ![grafik](https://user-images.githubusercontent.com/116741698/244881049-73e9a6ac-7be1-46fa-8d1b-854ae387b32c.png)

Yeah. This. More or less. Both together as sum. With 10s interval. (As I guess the inverter values are published every 5s), or both as stacked line diagram. So that you see the sum of both. 

In this view it looks like your meter reports -800W but your HM-1500 only produces 200W at 9.00. so you would have a total „consumption“ of -600W already. So if you don’t have several other inverters (I guess your single HM-600 won’t produce 600W at 9am, does it? And if it does, that means your total consumption would be 0W) your chart is wrong or misleading.

**Reply by OlSteph** [reply]
the problem with the powermeter values here - they have steps each 1 minute only, therfore i cannot go to 10 sec.
This is the other inverter - the HM-600:
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/3f4c6187-df3a-45b8-aa57-566c43f94d28)
around 9 at 340W

The HM-1500 around 9:00 was around 600W
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/450ae64c-f7b7-4f24-b295-8ac1f5ab06f4)

The powermeter said -814W at that time
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/116741698/1be72ec8-7ed7-47b8-802c-92602276ea4e)

This should mean than an this moment the consumption was 600+340-814=126Watts (normal power usage here if nothing is turned on like TV, PC etc)

**Reply by spcqike** [reply]
yes, and thats what i meant.

you have ~126W real consumption. in addition you produce ~344W with your HM-600.

this should result in a grid meter reading of -218W. assuming, that there are small timing differences and some difference while reading and calculating, it is possible that your HM-1500 got a grid meter reading of slightly under (or over?) -200W, so -195W or something. (in this particular timestamp: 9.00 am)

your old setting was target -200W, hysterese 200W. 

if i am right, and the hysterese is used wrongly, your inverter wont change its powerlimit target if your grid meter reading is inbetween -hysterese < 0 < hysterese. thats what you showed in your screenshot "reusing old limit" https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/256#discussioncomment-6123762) 

thats why i asked, if you can plot your "real" consumption ("real", because it must be the allready reduced consumption because of your always producing HM-600). i guess, its always within -200<0<200 W more or less (small timing differences, like mentioned above can occur. but the total meter reading shound't be <-200 or >200 for like 20-30 sec, otherwise it should have re-set the powerlimit of your HM-1500)

the relevant part in the code is to be found at
https://github.com/helgeerbe/OpenDTU-OnBattery/blob/master/src/PowerLimiter.cpp#L271

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/24735e62-7c21-4df1-8a6e-0e559af21775)


so, i guess thats a bug that happens, when the inverter is behind the powermeter. 

i also see, that the hysterese is ignored. at least in the way, i thought it would work.

i thought, with a target of -20W and a hystrese of 20W, it would reuse the old limit as long as the powermeter reports a value between -40W to 0W. but, my powerlimit is changed every 10s, even when the powermeter only changes by 1W 

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/79cad628-028d-41cb-881f-c6bcba69a0f0)

![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/836c4b41-b6c2-44c4-b4da-c175559d4b11)

**Reply by helgeerbe** [reply]
> i thought, with a target of -20W and a hystrese of 20W, it would reuse the old limit as long as the powermeter reports a value between -40W to 0W. but, my powerlimit is changed every 10s, even when the powermeter only changes by 1W

This seems indeed to be a bug. I will check this.

**Reply by spcqike** [reply]
i also tried to wrap my head around this. 

unfortunately i don't know how to create two PRs, as my first is still open :) so if i also change the hysterese, this PR would include the changed powerlimiter. How can one create a second, independent PR?

my solution would be something like store the newPowerLimit as a newly created _lastPowerLimit (right before return newPowerLimit) and check if newPowerLimit - _lastPowerLimit is within hysterese.

edit:
in my fork i simply changed acPower to int32_t and keep it out of the if(), so one can later check if(newPowerLimit-acPower) is out of hysterese, which seems to work.

```
int32_t acPower = 0;
    if (config.PowerLimiter_IsInverterBehindPowerMeter) {
        // If the inverter the behind the power meter (part of measurement),
        // the produced power of this inverter has also to be taken into account.
        // We don't use FLD_PAC from the statistics, because that
        // data might be too old and unreliable.
        acPower = static_cast<int>(inverter->Statistics()->getChannelFieldValue(TYPE_AC, (ChannelNum_t) config.PowerLimiter_InverterChannelId, FLD_PAC));
        newPowerLimit += acPower;
    }

    // We're not trying to hit 0 exactly but take an offset into account
    // This means we never fully compensate the used power with the inverter 
    // Case 3
    newPowerLimit -= config.PowerLimiter_TargetPowerConsumption;

    // Check if the new value is within the limits of the hysteresis and
    // if we can discharge the battery
    // If things did not change much we just use the old setting
    if ((newPowerLimit-acPower) >= (-config.PowerLimiter_TargetPowerConsumptionHysteresis) &&  
        (newPowerLimit-acPower) <= (+config.PowerLimiter_TargetPowerConsumptionHysteresis) &&
        batteryDischargeEnabled ) {
            MessageOutput.println("[PowerLimiterClass::loop] reusing old limit");
            return _lastRequestedPowerLimit;
    }
```

but of course only when target is reached.
![grafik](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/50098985/96807fe5-2e3b-4882-9232-ccc073d4ba6c)

in 1 and 2, there is a lot of changing, even so the target is reached, after the change, in 3, it uses the same PowerLimit value without changes.

**Reply by helgeerbe** [reply]
had the same idea. Will push it into development  in a minute.

**Comment by helgeerbe** [comment, marked as answer]
Should be fixed with [2023.06.12](https://github.com/helgeerbe/OpenDTU-OnBattery/releases/tag/2023.06.12)
